### PR TITLE
Make opencensus Stackdriver exporter respects initial_metadata option

### PR DESCRIPTION
**Pull request by bianpengyuan**
Currently Opencensus tracer uses [grpcService](https://github.com/envoyproxy/envoy/blob/4325949734fa10e4725bab5e096aa77cc8f85b06/generated_api_shadow/envoy/config/trace/v3/opencensus.proto#L72) proto to configure its [tracing client stub](https://github.com/envoyproxy/envoy/blob/e5366cc3d87c0a6d9e8b112b7f01abbcb6241b8f/source/extensions/tracers/opencensus/opencensus_tracer_impl.cc#L272). It uses GoogleGrpcUtil to construct a channel and create client stub with that. However, the channel created there does not take care of `initial_metadata` from grpcService configure. This change fills in OC`prepare_client_context` option in export to make it respect initial metadata.

Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by bianpengyuan** [comment]
@asraa I am having a hard time to find an easy way to test this. All current gRPC client tests are based on Envoy's async client/stream interface and implementation, but for this change we'd just want the test to verify the channel created could have initial metadata injected. Current options that I've thought about:
1) create a [GoogleAsyncClientImpl](https://github.com/envoyproxy/envoy/blob/9ad964db67b4d7bc9492b3cc097ba6f933e797af/source/common/grpc/google_async_client_impl.h#L165) with the initial metadata interceptor enabled channel, but leave `initial_metadata` in the [configuration param](https://github.com/envoyproxy/envoy/blob/9ad964db67b4d7bc9492b3cc097ba6f933e797af/source/common/grpc/google_async_client_impl.h#L169), and verify the stream could still have initial metadata injected.
2) Enable mock interface generation for [Greeter server](https://github.com/envoyproxy/envoy/blob/c66238a092b4b5c43c9661fa47bcd02d62f157bc/test/proto/helloworld.proto#L40) and test the generated channel with mock grpc client and server.

Neither seems ideal to me.. any better ideas?

**Comment by asraa** [comment]
re-assigning to @htuch,who may be able to help a little more with testing. 
I think maybe we should reduce the scope down to testing the written interceptor, but I'm uncertain how to do that right now.

**Comment by htuch** [comment]
@bianpengyuan a couple of things:
1. Can you explain in the commit message a bit more on the "what" vs. the motivation. I.e. we had some initial metadata plumbing before, what is the interceptor doing that provides value here? Ideally include links back to the Google gRPC docs of interest and how this would apply to end users.
2. We need some tests as discussed above. I think a good candidate place to put these would be https://github.com/envoyproxy/envoy/blob/master/test/common/grpc/google_async_client_impl_test.cc.

**Comment by bianpengyuan** [comment]
@htuch I updated the PR description. Yeah I have read through related tests. They are all based on AsyncClient / AsyncStream implementation and I don't find an easy way to apply them to this use case, which leads me to two options listed in https://github.com/envoyproxy/envoy/pull/11831#issuecomment-655236085.

**Comment by htuch** [comment]
@bianpengyuan looking at the description, this makes me think that we're doing something pretty weird in OpenCensus. We already have another thread about OpenCensus doing its own thing with HTTP and `curl` (https://github.com/envoyproxy/envoy/issues/11816), this is another place that it would be better for OpenCensus to provide a generic gRPC interface rather than relying on using Google gRPC libraries directly; when you do this you basically break Envoy's siloed threading model and lose observability via Envoy's normal mechanisms.

^^ is not actionable in this PR, but it would be good to make some progress here asynchronously, CC @g-easy @markdroth 

For tests, I think you're going to need to do an integration test with OpenCensus and gRPC tracing enabled with a fake upstream (it can be Envoy's inbuilt `FakeUpstream`) that can validate the initial metadata received.

**Comment by bianpengyuan** [comment]
@htuch find an easier way to achieve this with the lastest opencensus. ptal, thanks!

File: test/extensions/tracers/opencensus/config_test.cc [hunk 1]
```diff
@@ -322,6 +322,9 @@ TEST(OpenCensusTracerConfigTest, OpenCensusHttpTracerStackdriverGrpc) {
         google_grpc:
           target_uri: 127.0.0.1:55678
           stat_prefix: test
+        initial_metadata:
+        - key: foo
+          value: bar
```
Comment (htuch) [inline test/extensions/tracers/opencensus/config_test.cc:327]: Looks much nicer! Is there a way to verify that the options have been set in the OpenCensus registration?

Comment (bianpengyuan) [reply test/extensions/tracers/opencensus/config_test.cc:327]: I don't think so, It does not expose the client context or the stub.

Comment (htuch) [reply test/extensions/tracers/opencensus/config_test.cc:327]: OK, so this isn't any worse than the existing state of these tests. Before merging I'd like to check with @g-easy if there is a way or a plan to make this more testable that we can track.

Comment (g-easy) [reply test/extensions/tracers/opencensus/config_test.cc:327]: I'm sorry, I'm very limited in the time I can spend on OpenCensus. I can review PRs. How would you suggest improving testability here? In many cases, we limited our API on purpose but we can open it more where necessary.

Comment (htuch) [reply test/extensions/tracers/opencensus/config_test.cc:327]: @g-easy not asking for any development work, just advice on how we could construct an issue to track tech debt around testing? I.e. if someone had cycles, how would you advise they go about making options like this testable?

Comment (g-easy) [reply test/extensions/tracers/opencensus/config_test.cc:327]: For something like this, could we pass a mock stub to the Stackdriver trace exporter and expect that the extra metadata is included in requests?

There's already an option for passing custom stubs in the trace exporter options.

Comment (bianpengyuan) [reply test/extensions/tracers/opencensus/config_test.cc:327]: Yeah, that is feasible, but would require generating mock stub interface for stackdriver tracing service, and I am not sure whether Envoy would like to take that complexity or not.

Comment (htuch) [reply test/extensions/tracers/opencensus/config_test.cc:327]: @bianpengyuan is this just a simple gmock class? Or are you asking to engineer a fake object with some complexity? We'd definitely take the former.

Comment (bianpengyuan) [reply test/extensions/tracers/opencensus/config_test.cc:327]: just a gmock class, like https://grpc.github.io/grpc/cpp/md_doc_unit_testing.html

Comment (bianpengyuan) [reply test/extensions/tracers/opencensus/config_test.cc:327]: If that sounds good, I can add a TODO here to track testing for this.

Comment (htuch) [reply test/extensions/tracers/opencensus/config_test.cc:327]: Yes, thanks.

Comment (g-easy) [reply test/extensions/tracers/opencensus/config_test.cc:328]: Some time ago I persuaded the googleapis people to enable generate_mocks e.g. https://github.com/googleapis/googleapis/blob/5c8ab2c072d557c2f4c4e54b544394e2d62202d5/google/devtools/cloudtrace/v2/BUILD.bazel#L152

**Pull request merged**